crates/oxpinyin-python/tests_py/test_engine.py

```
@@ -0,0 +1,201 @@
+        # the native offsets are unsigned; a negative does not fit
+        engine.select(-1)
+
+
+def test_learning_requires_a_user_dir(make_engine):
+    engine = make_engine()  # no user_dir
+    with pytest.raises(OxpinyinError):
+        engine.train()
+    assert engine.save() is False
+
+
+def test_train_and_save_persist_user_state(make_engine, tmp_path, fixture_w3):
```

Ruff (0.16.2)

[warning] Unused function argument: `make_engine`

(ARG001)